### PR TITLE
Added Karma runner and BrowserStack testing

### PR DESCRIPTION
**Pull request by thelambdaparty**
PR for #16 

#### Karma Runner
[Karma](https://karma-runner.github.io/) facilitates the configuration for  adding extra services to the testing process. In this case adding BrowserStack testing was just a matter of installing [karma-browserstack-launcher](https://github.com/karma-runner/karma-browserstack-launcher) and adding [karma.browserstack.conf.ts](https://github.com/thelambdaparty/fetch-engine/blob/master/testing/karma.browserstack.conf.ts)

#### BrowserStack
karma-browserstack-launcher expects the following environment variables to be set for BrowserStack authentication: `BROWSERSTACK_USERNAME` and `BROWSERSTACK_KEY`.

These can be acquired in the BrowserStack [account settings page](https://www.browserstack.com/accounts/settings) at the bottom **Automate** section:

browserstack_automate

And added in the Travis repo settings page at the **Environment Variables** section:

travis_environment_vars

BrowserStack [Browsers & Mobile Devices for JavaScript Testing](https://www.browserstack.com/list-of-browsers-and-platforms?product=js_testing)

#### Testing Configuration
The files for testing configuration are in `./testing/`. All in TypeScript for linting, and compiled by the npm script `build-conf` (I haven't found a way to pipe tsc directly to karma without creating files).

### CODE REVIEW

**Comment by thelambdaparty** [comment]
Don't know if you're planning on semantic-release or any pull request formatting, so I tried to make it detailed and clear.

File: package.json [hunk 1]
```diff
@@ -4,19 +4,32 @@
...
   "main": "dist/index.js",
   "scripts": {
-    "clean": "rm -rf ./dist",
-    "build": "npm run clean && tsc",
+    "//": "Typescript Compilation",
+
+    "//": "`./testing` files are in typescript for linting, ",
+    "//": "`build-conf`: compiles them for running with karma.",
+
+    "//": "the `|| true` is so that typescript doesn't",
```
Comment (phuu) [inline package.json]: TypeScript's checks need to fail the build — they usually mean something is actually broken. Was there something specific that was failing unexpectedly?

File: package.json [hunk 2]
```diff
@@ -4,19 +4,32 @@
   "description": "A smart request-making library that makes sure your Javascript client is a good citizen of your distributed system.",
   "main": "dist/index.js",
   "scripts": {
-    "clean": "rm -rf ./dist",
-    "build": "npm run clean && tsc",
+    "//": "Typescript Compilation",
```
Comment (phuu) [inline package.json]: In general, I'm not a huge fan of docs like this. The scripts are simple enough to be self explanatory – more docs should go in the README.md

**Comment by phuu** [comment]
Hey @thelambdaparty — thanks for this. I was a bit unclear in the issue originally (sorry about that) — it's totally fine for the `testing` files to be JavaScript, I just meant it'd be great to match code-style.

However, I'll probably merge this onto a branch and iterate until the build is passing. Sound good?

**Comment by thelambdaparty** [comment]
@phuu Sure! 

I added the comments there just so you know what I did, README or docs will be better. The `|| true` for Typescript was just for testing the rest of the build (see if it runs ok), but it should surely be removed, otherwise it's evading TS's best feature :smile: 

Another branch for iteration sounds cool :+1:

**Comment by phuu** [comment]
Good stuff.

GitHub can't change target branches for PRs — would you mind closing and reopening on something like a like `karma` branch please? I'll merge and pickup getting it passing on travis.

**Comment by thelambdaparty** [comment]
Created branch karma: https://github.com/thelambdaparty/fetch-engine/tree/karma
Will you open a karma branch too or send PR to browserstack branch?

**Comment by phuu** [comment]
@thelambdaparty I've just set up a karma branch that you should be able to create a PR for. Thanks :)

**Comment by phuu** [comment]
@thelambdaparty https://github.com/phuu/fetch-engine/pull/54

**Pull request closed without merging**